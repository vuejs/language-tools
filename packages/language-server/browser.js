module.exports = require('./out/browser');
