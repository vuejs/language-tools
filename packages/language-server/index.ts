export * from './lib/types';
export * from './lib/initialize';

// export types of parent package
export * from '@volar/language-server/lib/types';

// only export types of depend packages
export * from '@vue/language-service/lib/types';
