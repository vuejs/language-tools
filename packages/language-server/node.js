module.exports = require('./out/node');
