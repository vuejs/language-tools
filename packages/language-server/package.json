{
	"name": "@vue/language-server",
	"version": "2.0.15",
	"license": "MIT",
	"files": [
		"**/*.js",
		"**/*.d.ts"
	],
	"bin": {
		"vue-language-server": "./bin/vue-language-server.js"
	},
	"repository": {
		"type": "git",
		"url": "https://github.com/vuejs/language-tools.git",
		"directory": "packages/language-server"
	},
	"dependencies": {
		"@volar/language-core": "2.2.0-alpha.12",
		"@volar/language-server": "2.2.0-alpha.12",
		"@vue/language-core": "2.0.15",
		"@vue/language-service": "2.0.15",
		"@vue/typescript-plugin": "2.0.15",
		"vscode-languageserver-protocol": "^3.17.5"
	}
}
