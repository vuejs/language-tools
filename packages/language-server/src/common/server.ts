import * as vscode from 'vscode-languageserver';
import { URI } from 'vscode-uri';
import { FileSystemHost, LanguageServerInitializationOptions, LanguageServerPlugin, RuntimeEnvironment, ServerMode } from '../types';
import { createCancellationTokenHost } from './cancellationPipe';
import { createConfigurationHost } from './configurationHost';
import { createDocuments } from './documents';
import { setupCapabilities } from './utils/registerFeatures';
import { createWorkspaces } from './workspaces';

export interface ServerContext {
	connection: vscode.Connection,
	runtimeEnv: RuntimeEnvironment,
	plugins: LanguageServerPlugin[],
}

export function createCommonLanguageServer(context: ServerContext) {

	let initParams: vscode.InitializeParams;
	let options: LanguageServerInitializationOptions;
	let roots: URI[] = [];
	let fsHost: FileSystemHost | undefined;
	let projects: ReturnType<typeof createWorkspaces> | undefined;
	let configurationHost: ReturnType<typeof createConfigurationHost> | undefined;
	let plugins: ReturnType<LanguageServerPlugin>[];

	const documents = createDocuments(context.connection);

	context.connection.onInitialize(async _params => {

		initParams = _params;
		options = initParams.initializationOptions;
		plugins = context.plugins.map(plugin => plugin(options));

		if (initParams.capabilities.workspace?.workspaceFolders && initParams.workspaceFolders) {
			roots = initParams.workspaceFolders.map(folder => URI.parse(folder.uri));
		}
		else if (initParams.rootUri) {
			roots = [URI.parse(initParams.rootUri)];
		}
		else if (initParams.rootPath) {
			roots = [URI.file(initParams.rootPath)];
		}

		const result: vscode.InitializeResult = {
			capabilities: {
				textDocumentSync: (options.textDocumentSync as vscode.TextDocumentSyncKind) ?? vscode.TextDocumentSyncKind.Incremental,
			},
		};

		configurationHost = initParams.capabilities.workspace?.configuration ? createConfigurationHost(initParams, context.connection) : undefined;

		setupCapabilities(initParams.capabilities, result.capabilities, options, plugins, getSemanticTokensLegend());
		await createLanguageServiceHost();

		try {
			// show version on LSP logs
			const packageJson = require('../package.json');
			result.serverInfo = {
				name: packageJson.name,
				version: packageJson.version,
			};
		} catch { }

		return result;
	});
	context.connection.onInitialized(() => {

		fsHost?.ready(context.connection);
		configurationHost?.ready();

		if (initParams.capabilities.workspace?.workspaceFolders) {
			context.connection.workspace.onDidChangeWorkspaceFolders(e => {

				for (const folder of e.added) {
					projects?.add(URI.parse(folder.uri));
				}

				for (const folder of e.removed) {
					projects?.remove(URI.parse(folder.uri));
				}
			});
		}

		if (
			options.serverMode !== ServerMode.Syntactic
			&& !options.disableFileWatcher
			&& initParams.capabilities.workspace?.didChangeWatchedFiles?.dynamicRegistration
		) {
			context.connection.client.register(vscode.DidChangeWatchedFilesNotification.type, {
				watchers: [
					{
						globPattern: `**/*.{${[
							'js',
							'cjs',
							'mjs',
							'ts',
							'cts',
							'mts',
							'jsx',
							'tsx',
							'json',
							...plugins.map(plugin => plugin.extraFileExtensions.map(ext => ext.extension)).flat(),
						].join(',')}}`
					},
				]
			});
		}
	});
	context.connection.listen();

	async function createLanguageServiceHost() {

		const ts = context.runtimeEnv.loadTypescript(options.typescript.tsdk);
		fsHost = context.runtimeEnv.createFileSystemHost(ts, initParams.capabilities);

		const tsLocalized = initParams.locale ? await context.runtimeEnv.loadTypescriptLocalized(options.typescript.tsdk, initParams.locale) : undefined;
		const cancelTokenHost = createCancellationTokenHost(options.cancellationPipeName);
		const _projects = createWorkspaces({
			server: context,
			fileSystemHost: fsHost,
			configurationHost,
			ts,
			tsLocalized,
			initParams: initParams,
			initOptions: options,
			documents,
			cancelTokenHost,
			plugins,
		});
		projects = _projects;

		for (const root of roots) {
			projects.add(root);
		}

		(await import('./features/customFeatures')).register(context.connection, projects);
		(await import('./features/languageFeatures')).register(context.connection, projects, initParams, cancelTokenHost, getSemanticTokensLegend());

		for (const plugin of plugins) {
			plugin.onInitialize?.(context.connection, getLanguageService as any);
		}

		async function getLanguageService(uri: string) {
			const project = (await projects!.getProject(uri))?.project;
			return project?.getLanguageService();
		}
	}

	function getSemanticTokensLegend() {
		if (!options.semanticTokensLegend) {
			return standardSemanticTokensLegend;
		}
		return {
			tokenTypes: [...standardSemanticTokensLegend.tokenTypes, ...options.semanticTokensLegend.tokenTypes],
			tokenModifiers: [...standardSemanticTokensLegend.tokenModifiers, ...options.semanticTokensLegend.tokenModifiers],
		};
	}
}

// https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#standard-token-types-and-modifiers
const standardSemanticTokensLegend: vscode.SemanticTokensLegend = {
	tokenTypes: [
		'namespace',
		'class',
		'enum',
		'interface',
		'struct',
		'typeParameter',
		'type',
		'parameter',
		'variable',
		'property',
		'enumMember',
		'decorator',
		'event',
		'function',
		'method',
		'macro',
		'label',
		'comment',
		'string',
		'keyword',
		'number',
		'regexp',
		'operator',
	],
	tokenModifiers: [
		'declaration',
		'definition',
		'readonly',
		'static',
		'deprecated',
		'abstract',
		'async',
		'modification',
		'documentation',
		'defaultLibrary',
	],
};
