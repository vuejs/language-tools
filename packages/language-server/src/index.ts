export * from './requests';
export * from './types';
