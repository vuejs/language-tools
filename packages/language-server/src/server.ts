import * as vscode from 'vscode-languageserver';
import { URI } from 'vscode-uri';
import { FileSystemHost, LanguageServerPlugin, RuntimeEnvironment, ServerInitializationOptions } from './types';
import { createConfigurationHost } from './utils/configurationHost';
import { createDocumentServiceHost } from './utils/documentServiceHost';
import { createSnapshots } from './utils/snapshots';
import { createWorkspaces } from './utils/workspaces';

export function createCommonLanguageServer(
	connection: vscode.Connection,
	runtimeEnv: RuntimeEnvironment,
	_plugins: LanguageServerPlugin[],
) {

	let params: vscode.InitializeParams;
	let options: ServerInitializationOptions;
	let roots: URI[] = [];
	let fsHost: FileSystemHost | undefined;
	let projects: ReturnType<typeof createWorkspaces> | undefined;
	let documentServiceHost: ReturnType<typeof createDocumentServiceHost> | undefined;
	let configHost: ReturnType<typeof createConfigurationHost> | undefined;
	let plugins: ReturnType<LanguageServerPlugin>[];

	const documents = createSnapshots(connection);

	connection.onInitialize(async _params => {

		params = _params;
		options = params.initializationOptions as any;
		plugins = _plugins.map(plugin => plugin(options));

		if (params.capabilities.workspace?.workspaceFolders && params.workspaceFolders) {
			roots = params.workspaceFolders.map(folder => URI.parse(folder.uri));
		}
		else if (params.rootUri) {
			roots = [URI.parse(params.rootUri)];
		}
		else if (params.rootPath) {
			roots = [URI.file(params.rootPath)];
		}

		const result: vscode.InitializeResult = {
			capabilities: {
				textDocumentSync: (options.textDocumentSync as vscode.TextDocumentSyncKind) ?? vscode.TextDocumentSyncKind.Incremental,
			},
		};
		const ts = runtimeEnv.loadTypescript(options);

		configHost = params.capabilities.workspace?.configuration ? createConfigurationHost(params, connection) : undefined;

		if (options.documentFeatures) {

			(await import('./registers/registerDocumentFeatures')).register(options.documentFeatures, result.capabilities);

			documentServiceHost = createDocumentServiceHost(
				runtimeEnv,
				plugins,
				ts,
				configHost,
			);

			for (const root of roots) {
				documentServiceHost.add(root);
			}

			(await import('./features/documentFeatures')).register(connection, documents, documentServiceHost);
		}
		if (options.languageFeatures) {
			(await import('./registers/registerlanguageFeatures')).register(options.languageFeatures!, result.capabilities, plugins);

			fsHost = runtimeEnv.createFileSystemHost(ts, params.capabilities);

			const tsLocalized = runtimeEnv.loadTypescriptLocalized(options);

			projects = createWorkspaces(
				runtimeEnv,
				plugins,
				fsHost,
				configHost,
				ts,
				tsLocalized,
				options,
				documents,
				connection,
			);

			for (const root of roots) {
				projects.add(root);
			}

			(await import('./features/customFeatures')).register(connection, projects, plugins);
			(await import('./features/languageFeatures')).register(connection, projects, options.languageFeatures, params);
		}

		try {
			// show version on LSP logs
			const packageJson = require('../package.json');
			result.serverInfo = {
				name: packageJson.name,
				version: packageJson.version,
			};
		} catch { }

		return result;
	});
	connection.onInitialized(async () => {

		fsHost?.ready(connection);
		configHost?.ready();

		if (params.capabilities.workspace?.workspaceFolders) {
			connection.workspace.onDidChangeWorkspaceFolders(e => {

				for (const folder of e.added) {
					documentServiceHost?.add(URI.parse(folder.uri));
					projects?.add(URI.parse(folder.uri));
				}

				for (const folder of e.removed) {
					documentServiceHost?.remove(URI.parse(folder.uri));
					projects?.remove(URI.parse(folder.uri));
				}
			});
		}
	});
	connection.listen();
}
