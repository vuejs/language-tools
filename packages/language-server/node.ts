import type { Connection } from '@volar/language-server';
import { createConnection, createServer, createTypeScriptProjectProvider, loadTsdkByPath } from '@volar/language-server/node';
import { ParsedCommandLine, VueCompilerOptions, createParsedCommandLine, createVueLanguagePlugin, parse, resolveVueCompilerOptions } from '@vue/language-core';
import { ServiceEnvironment, convertAttrName, convertTagName, createDefaultGetTsPluginClient, detect, getVueLanguageServicePlugins } from '@vue/language-service';
import * as tsPluginClient from '@vue/typescript-plugin/lib/client';
import { searchNamedPipeServerForFile } from '@vue/typescript-plugin/lib/utils';
import { GetLanguagePlugin, createHybridModeProjectProviderFactory } from './lib/hybridModeProject';
import { DetectNameCasingRequest, GetConnectedNamedPipeServerRequest, GetConvertAttrCasingEditsRequest, GetConvertTagCasingEditsRequest, ParseSFCRequest } from './lib/protocol';
import type { VueInitializationOptions } from './lib/types';

let tsdk: ReturnType<typeof loadTsdkByPath>;
let hybridMode: boolean;
let getTsPluginClient: ReturnType<typeof createDefaultGetTsPluginClient>;

const envToVueOptions = new WeakMap<ServiceEnvironment, VueCompilerOptions>();
const watchedExtensions = new Set<string>();

export const connection: Connection = createConnection();

export const server = createServer(connection);

export const getLanguagePlugins: GetLanguagePlugin = async (serviceEnv, configFileName, host, sys) => {
	const commandLine = await parseCommandLine();
	const vueOptions = commandLine?.vueOptions ?? resolveVueCompilerOptions({});
	const vueLanguagePlugin = createVueLanguagePlugin(
		tsdk.typescript,
		serviceEnv.typescript!.uriToFileName,
		sys?.useCaseSensitiveFileNames ?? false,
		() => host?.getProjectVersion?.() ?? '',
		() => host?.getScriptFileNames() ?? [],
		commandLine?.options ?? {},
		vueOptions,
	);
	if (!hybridMode) {
		const extensions = [
			'js', 'cjs', 'mjs', 'ts', 'cts', 'mts', 'jsx', 'tsx', 'json',
			...vueOptions.extensions.map(ext => ext.slice(1)),
			...vueOptions.vitePressExtensions.map(ext => ext.slice(1)),
			...vueOptions.petiteVueExtensions.map(ext => ext.slice(1)),
		];
		const newExtensions = extensions.filter(ext => !watchedExtensions.has(ext));
		if (newExtensions.length) {
			for (const ext of newExtensions) {
				watchedExtensions.add(ext);
			}
			server.watchFiles(['**/*.{' + newExtensions.join(',') + '}']);
		}
	}

	envToVueOptions.set(serviceEnv, vueOptions);

	return [vueLanguagePlugin];

	async function parseCommandLine() {
		let commandLine: ParsedCommandLine | undefined;
		let sysVersion: number | undefined;
		if (sys) {
			let newSysVersion = await sys.sync();
			while (sysVersion !== newSysVersion) {
				sysVersion = newSysVersion;
				if (configFileName) {
					commandLine = createParsedCommandLine(tsdk.typescript, sys, configFileName);
				}
				newSysVersion = await sys.sync();
			}
		}
		return commandLine;
	}
};

connection.listen();

connection.onInitialize(params => {
	const options: VueInitializationOptions = params.initializationOptions;

	hybridMode = options.vue?.hybridMode ?? true;
	tsdk = loadTsdkByPath(options.typescript.tsdk, params.locale);

	if (hybridMode) {
		getTsPluginClient = () => tsPluginClient;
	}
	else {
		getTsPluginClient = createDefaultGetTsPluginClient(tsdk.typescript, env => envToVueOptions.get(env)!);
	}

	const result = server.initialize(
		params,
		getVueLanguageServicePlugins(
			tsdk.typescript,
			env => envToVueOptions.get(env)!,
			getTsPluginClient,
			hybridMode,
		),
		hybridMode
			? createHybridModeProjectProviderFactory(tsdk.typescript.sys, getLanguagePlugins)
			: createTypeScriptProjectProvider(tsdk.typescript, tsdk.diagnosticMessages, (env, ctx) => getLanguagePlugins(env, ctx.configFileName, ctx.host, ctx.sys)),
		{
			pullModelDiagnostics: hybridMode,
		},
	);
	if (hybridMode) {
		// provided by tsserver + @vue/typescript-plugin
		result.capabilities.semanticTokensProvider = undefined;
	}
	return result;
});

connection.onInitialized(() => {
	server.initialized();
});

connection.onShutdown(() => {
	server.shutdown();
});

connection.onRequest(ParseSFCRequest.type, params => {
	return parse(params);
});

connection.onRequest(DetectNameCasingRequest.type, async params => {
	const languageService = await getService(params.textDocument.uri);
	if (languageService) {
		return await detect(languageService.context, params.textDocument.uri);
	}
});

connection.onRequest(GetConvertTagCasingEditsRequest.type, async params => {
	const languageService = await getService(params.textDocument.uri);
	if (languageService) {
		return await convertTagName(languageService.context, params.textDocument.uri, params.casing, getTsPluginClient(languageService.context));
	}
});

connection.onRequest(GetConvertAttrCasingEditsRequest.type, async params => {
	const languageService = await getService(params.textDocument.uri);
	if (languageService) {
		return await convertAttrName(languageService.context, params.textDocument.uri, params.casing, getTsPluginClient(languageService.context));
	}
});

connection.onRequest(GetConnectedNamedPipeServerRequest.type, async fileName => {
	const server = (await searchNamedPipeServerForFile(fileName))?.server;
	if (server) {
		return server;
	}
});

async function getService(uri: string) {
	return (await server.projects.get.call(server, uri)).getLanguageService();
}
