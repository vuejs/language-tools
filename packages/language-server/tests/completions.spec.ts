import { InsertReplaceEdit, TextDocument, TextEdit } from '@volar/language-server';
import { afterEach, describe, expect, it } from 'vitest';
import { URI } from 'vscode-uri';
import { getLanguageServer, testWorkspacePath } from './server.js';

describe('Completions', async () => {

	it('Directives', async () => {
		await assertCompletion('fixture.vue', 'vue', `<template><div v-ht|></div></template>`, 'v-html');
		await assertCompletion('fixture.vue', 'vue', `<template><div v-cl|></div></template>`, 'v-cloak');
		await assertCompletion('fixture.vue', 'vue', `<template><div v-el|></div></template>`, 'v-else');
		await assertCompletion('fixture.vue', 'vue', `<template><div v-p|></div></template>`, 'v-pre');
	});

	it('$event argument', async () => {
		await assertCompletion('fixture.vue', 'vue', `<template><div @click="console.log($eve|)"></div></template>`, '$event');
	});

	it('<script setup>', async () => {
		await assertCompletion('fixture.vue', 'vue', `
			<template>{{ f| }}</template>

			<script lang="ts" setup>
			const foo = 1;
			</script>
		`, 'foo');
	});

	it('Slot name', async () => {
		await assertCompletion('fixture.vue', 'vue', `
			<template>
				<Foo>
					<template #|></template>
				</Foo>
			</template>

			<script lang="ts" setup>
			let Foo: new () => {
				$slots: {
					default: any;
				};
			};
			</script>
		`, 'default');
	});

	it('#2454', async () => {
		await assertCompletion('fixture.vue', 'vue', `
			<script setup lang="ts">
			let vLoading: any;
			</script>

			<template>
			<div v-load|="vLoading"></div>
			</template>
		`, 'v-loading');
	});

	it('#2511', async () => {
		await ensureGlobalTypesHolder('tsconfigProject');
		await openDocument('tsconfigProject/component-for-auto-import.vue', 'vue', `<script setup lang="ts"></script>`);
		await assertCompletion('tsconfigProject/fixture.vue', 'vue', `
			<script setup lang="ts">
			import componentFor|
			</script>
		`, 'ComponentForAutoImport');
	});

	it('#3658', async () => {
		await assertCompletion('fixture.vue', 'vue', `
			<template>
				<Comp>
					<template #foo="foo">
						{{ fo| }}
					</template>
				</Comp>
			</template>
		`, 'foo');
	});

	it('#4639', async () => {
		await assertCompletion('fixture.vue', 'vue', `
			<template>
				<div @click.| />
			</template>
		`, 'capture');
	});

	it('Alias path', async () => {
		await ensureGlobalTypesHolder('tsconfigProject');
		await assertCompletion('tsconfigProject/fixture.vue', 'vue', `
			<script setup lang="ts">
			import Component from '@/|';
			</script>
		`, 'empty.vue');
	});

	it('Relative path', async () => {
		await ensureGlobalTypesHolder('tsconfigProject');
		await assertCompletion('tsconfigProject/fixture.vue', 'vue', `
			<script setup lang="ts">
			import Component from './|';
			</script>
		`, 'empty.vue');
	});

	it('Component auto import', async () => {
		await ensureGlobalTypesHolder('tsconfigProject');
		await openDocument('tsconfigProject/ComponentForAutoImport.vue', 'vue', `<script setup lang="ts"></script>`);
		await assertCompletion('tsconfigProject/fixture.vue', 'vue', `
			<script setup lang="ts">
			</script>

			<template>
				<ComponentForA| />
			</template>
		`, 'ComponentForAutoImport');
	});

	it('core#8811', async () => {
		await ensureGlobalTypesHolder('tsconfigProject');
		await assertCompletion('tsconfigProject/fixture.vue', 'vue', `
			<script setup lang="ts">
			declare const Foo: new () => {
				$props: {
					FooBar: string;
				};
			};
			</script>

			<template>
				<Foo :-| ></Foo>
			</template>
		`, ':-foo-bar');
	});

	const openedDocuments: TextDocument[] = [];

	afterEach(async () => {
		const server = await getLanguageServer();
		for (const document of openedDocuments) {
			await server.closeTextDocument(document.uri);
		}
		openedDocuments.length = 0;
	});

	/**
	 * @deprecated Remove this when #4717 fixed.
	 */
	async function ensureGlobalTypesHolder(folderName: string) {
		const document = await openDocument(`${folderName}/globalTypesHolder.vue`, 'vue', '');
		const server = await getLanguageServer();
		await server.sendDocumentDiagnosticRequest(document.uri);
	}

	async function assertCompletion(fileName: string, languageId: string, content: string, itemLabel: string) {
		const offset = content.indexOf('|');
		content = content.slice(0, offset) + content.slice(offset + 1);

		const server = await getLanguageServer();
		let document = await openDocument(fileName, languageId, content);

		const position = document.positionAt(offset);
		const completions = await server.sendCompletionRequest(document.uri, position);

		let completion = completions?.items.find(item => item.label === itemLabel);
		expect(completion).toBeDefined();

		completion = await server.sendCompletionResolveRequest(completion!);
		expect(completion).toBeDefined();

		const edits: TextEdit[] = [];

		if (completion.textEdit) {
			if (InsertReplaceEdit.is(completion.textEdit)) {
				edits.push({ newText: completion.textEdit.newText, range: completion.textEdit.replace });
			}
			else {
				edits.push(completion.textEdit);
			}
		}
		else {
			edits.push({
				newText: completion.insertText ?? completion.label,
				range: { start: position, end: position },
			});
		}
		if (completion.additionalTextEdits) {
			edits.push(...completion.additionalTextEdits);
		}
		if (edits.length === 0) {
			console.log(completion);
		}
		expect(edits.length).toBeGreaterThan(0);

		document = await server.updateTextDocument(document.uri, edits);

		expect(document.getText()).toMatchSnapshot();
	}

	async function openDocument(fileName: string, languageId: string, content: string) {
		const server = await getLanguageServer();
		const uri = URI.file(`${testWorkspacePath}/${fileName}`);
		const document = await server.openInMemoryDocument(uri.toString(), languageId, content);
		if (openedDocuments.every(d => d.uri !== document.uri)) {
			openedDocuments.push(document);
		}
		return document;
	}
});
