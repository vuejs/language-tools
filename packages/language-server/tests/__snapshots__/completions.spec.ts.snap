// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Completions > #2454 1`] = `
"
			<script setup lang="ts">
			let vLoading: any;
			</script>

			<template>
			<div v-loading="vLoading"></div>
			</template>
		"
`;

exports[`Completions > #2511 1`] = `
"
			<script setup lang="ts">
			import componentForAutoImport$1 from './component-for-auto-import.vue';
			</script>
		"
`;

exports[`Completions > #3658 1`] = `
"
			<template>
				<Comp>
					<template #foo="foo">
						{{ foo }}
					</template>
				</Comp>
			</template>
		"
`;

exports[`Completions > #4639 1`] = `
"
			<template>
				<div @click.capture />
			</template>
		"
`;

exports[`Completions > $event argument 1`] = `"<template><div @click="console.log($event)"></div></template>"`;

exports[`Completions > <script setup> 1`] = `
"
			<template>{{ foo }}</template>

			<script lang="ts" setup>
			const foo = 1;
			</script>
		"
`;

exports[`Completions > Alias path 1`] = `
"
			<script setup lang="ts">
			import Component from '@/empty.vue';
			</script>
		"
`;

exports[`Completions > Component auto import 1`] = `
"
			<script setup lang="ts">
import ComponentForAutoImport from './ComponentForAutoImport.vue';

			</script>

			<template>
				<ComponentForAutoImport />
			</template>
		"
`;

exports[`Completions > Directives 1`] = `"<template><div v-html="$1"></div></template>"`;

exports[`Completions > Directives 2`] = `"<template><div v-cloak></div></template>"`;

exports[`Completions > Directives 3`] = `"<template><div v-else></div></template>"`;

exports[`Completions > Directives 4`] = `"<template><div v-pre></div></template>"`;

exports[`Completions > Relative path 1`] = `
"
			<script setup lang="ts">
			import Component from './empty.vue';
			</script>
		"
`;

exports[`Completions > Slot name 1`] = `
"
			<template>
				<Foo>
					<template #default></template>
				</Foo>
			</template>

			<script lang="ts" setup>
			let Foo: new () => {
				$slots: {
					default: any;
				};
			};
			</script>
		"
`;

exports[`Completions > core#8811 1`] = `
"
			<script setup lang="ts">
			declare const Foo: new () => {
				$props: {
					FooBar: string;
				};
			};
			</script>

			<template>
				<Foo :-foo-bar="$1" ></Foo>
			</template>
		"
`;
